### PR TITLE
Improved thread safety

### PR DESCRIPTION
**Pull request by QuestofIranon**
Encountered data races and access violation crashes due to the cached date formatters array while using the library. I fixed it by wrapping it in a class and adding a dispatch queue/lock. 

I left the new class as an internal private subclass of the Date extension but it could be moved if desired. This may not fix all multi threading issues, but so far that is the only one I've encountered.

### CODE REVIEW

**Comment by QuestofIranon** [comment]
Might want to hold off on merging, I'm going to give the other cache the same treatment.

**Comment by QuestofIranon** [comment]
Should be fine now.

File: Sources/DateHelper.swift [hunk 1]
```diff
@@ -491,42 +495,91 @@ public extension Date {
...
-            formatter.timeZone = timeZone
-            formatter.locale = locale
-            formatter.isLenient = true
-            Date.cachedDateFormatters[hashKey] = formatter
+        
+        private func retrieve(hashKey: String) -> DateFormatter? {
+            let dateFormatter = concurrentFormatterCache.cachedDateFormattersQueue.sync { () -> DateFormatter? in
+                guard let result = concurrentFormatterCache.cachedDateFormatters[hashKey] else { return nil }
+                
+                return result.copy() as? DateFormatter
```
Comment (corban123) [inline Sources/DateHelper.swift]: @QuestofIranon Quick question, what's the point of this copy right here?

**Comment by weakfl** [comment]
@melvitax Any plans to merge this PR?

I'd like to submit a PR for Swift 5 support, but this one and #86 should probably be merged first and a new release for Swift 4.2 pushed to cocoapods?

File: Sources/DateHelper.swift (hunk 1 above)
Comment (QuestofIranon) [reply Sources/DateHelper.swift]: Sorry, I somehow missed this comment. I don't have access to a mac anymore, I suspect my concern was related to avoiding concurrent access to the result by copying it, but I can't tell if it was an actual problem to be concerned about.

**Pull request merged**